feat: add email commands

ColinFrick opened the pull request:
This PR adds the following email commands:
- `fusionauth email:create` - Create a new email template locally.
- `fusionauth email:download` - Download a specific template or all email templates from a FusionAuth server.
- `fusionauth email:duplicate` - Duplicate an email template locally.
- `fusionauth email:html-to-text` - Convert HTML email templates to text, where the text template is missing.
- `fusionauth email:upload` - Upload a specific template or all email templates to a FusionAuth server.
- `fusionauth email:watch` - Watch the email template directory and upload changes to a FusionAuth server.

The email directory is as follows:
```
emails
├── 004e5f2e-ec50-4eda-998e-c58afd2a048d
│   ├── body.html
│   ├── body.txt
│   ├── from_email.txt
│   ├── from_name.txt
│   ├── fuubar.txt
│   ├── name.txt
│   └── subject.txt
├── 530100f3-e25b-49e3-bdd0-46e4c2f06b9e
│   ├── de
│   │   ├── body.html
│   │   ├── body.txt
│   │   ├── from_name.txt
│   │   └── subject.txt
│   ├── fr
│   │   ├── body.html
│   │   ├── body.txt
│   │   ├── from_name.txt
│   │   └── subject.txt
│   ├── body.html
│   ├── body.txt
│   ├── from_email.txt
│   ├── from_name.txt
│   ├── name.txt
│   └── subject.txt
```

So we have a directory for each template containing the base properties and default templates. Locales are subdirectory containing the localized templates.

Closes #2

ColinFrick committed:
refactor: extract theme-specific options

ColinFrick committed:
feat: add email commands

mooreds commented:
@vcampitelli did you have a chance to look at this?

vcampitelli commented on src/commands/email-download.ts:
```diff
@@ -0,0 +1,107 @@
+import {Command} from "commander";
+import {getEmailErrorMessage, getEmailSuccessMessage, reportError, validateEmailOptions} from "../utils.js";
+import {EmailTemplate, FusionAuthClient} from "@fusionauth/typescript-client";
+import * as fs from "fs";
```
```suggestion
import {existsSync} from "fs";
```

vcampitelli commented on src/commands/email-download.ts:
```diff
@@ -0,0 +1,107 @@
+                const emailTemplateId = emailTemplate.id;
+                const emailTemplateDirectory = `${output}/${emailTemplateId}/`;
+
+                if (!fs.existsSync(emailTemplateDirectory)) {
```
```suggestion
                if (!existsSync(emailTemplateDirectory)) {
```

vcampitelli commented on src/commands/email-download.ts:
```diff
@@ -0,0 +1,107 @@
+                const locales = getLocalesFromEmailTemplates(emailTemplate);
+                for (const locale of locales) {
+                    const localeDirectory = `${emailTemplateDirectory}/${locale}/`;
+                    if (!fs.existsSync(localeDirectory)) {
```
```suggestion
                    if (!existsSync(localeDirectory)) {
```

vcampitelli commented on src/commands/email-download.ts:
```diff
@@ -0,0 +1,107 @@
+ * Gets the locales from a list of email templates
+ * @param emailTemplate
+ */
+export const getLocalesFromEmailTemplates = (emailTemplate: EmailTemplate): string[] => {
```
Nice :sunglasses:

vcampitelli reviewed (changes requested):
Just a few nitpicks. Great work though!

ColinFrick committed:
refactor: use @commander-js/extra-typings

This commit removes our own type definitions and utilizes '@commander-js/extra-typings' to organize the option-related types better. Because of this all custom validation code was rendered obsolete and was consequently removed.
The primary objective behind these changes is to keep the codebase more maintainable and efficient in terms of usage of external dependencies that are better designed for the task at hand.

ColinFrick committed:
Merge branch 'feat/typings' into feat/emails

# Conflicts:
#	README.md
#	package-lock.json
#	package.json
#	src/commands/index.ts
#	src/commands/theme-download.ts
#	src/commands/theme-upload.ts
#	src/commands/theme-watch.ts
#	src/index.ts
#	src/utils.ts

ColinFrick committed:
refactor: use @commander-js/extra-typings for email commands

mooreds commented:
@vcampitelli can you take another look, please?

vcampitelli committed:
:bug: Removing unwanted `export` that was causing an error

vcampitelli commented:
I pushed f1a38d94278af88ec41b8592a03947faa5ae761b to remove an `export` because the script was trying to add it as a command and it's just a helper function.

```
src/index.ts:23:63 - error TS2345: Argument of type 'Command<[], { output: string; locales: [] | string[]; }> | Command<[string | undefined], { output: string; }> | Command<[string | undefined], { output: string; key: string; host: string; clean: boolean; }> | ... 8 more ... | Command<...>' is not assignable to parameter of type 'CommandUnknownOpts'.
  Type '(emailTemplate: EmailTemplate) => string[]' is not assignable to type 'Command<unknown[], OptionValues>'.

23 Object.values(commands).forEach(command => program.addCommand(command));
```                                                                 ~~~~~~~
